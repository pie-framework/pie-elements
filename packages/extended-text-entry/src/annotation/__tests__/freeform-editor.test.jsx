import { shallow } from 'enzyme';
import React from 'react';
import FreeformEditor from '../freeform-editor';
import { unwrap } from "@material-ui/core/test-utils";

const Editor = unwrap(FreeformEditor);

describe('freeform editor', () => {
  const defaultProps = {
    classes: {},
    anchorEl: {},
    open: false,
    offset: 10,
    value: 'note',
    type: 'positive'
  };
  let onClose;
  let onDelete;
  let onSave;
  let onTypeChange;

  beforeEach(() => {
    onClose = jest.fn();
    onDelete = jest.fn();
    onSave = jest.fn();
    onTypeChange = jest.fn();
  });

  const wrapper = extras => {
    const defaults = {
      ...defaultProps,
      onClose,
      onDelete,
      onSave,
      onTypeChange
    };
    const props = { ...defaults, ...extras };

    return shallow(<Editor { ...props } />);
  };

  describe('snapshots', () => {
    it('renders', () => {
      expect(wrapper()).toMatchSnapshot();
    });

    it('opened', () => {
      expect(wrapper({ open: true })).toMatchSnapshot();
    });

    it('opened, negative type', () => {
      expect(
        wrapper({ open: true, type: 'negative' })
      ).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    describe('save or autosave', () => {
      it('calls onSave',  () => {
        const w = wrapper();

        w.setState({ value: 'good' });
        w.instance().handleSave();
        expect(onSave).toBeCalled();
      });

      it('calls onDelete', () => {
        const w = wrapper();

        w.setState({ value: '' });
        w.instance().handleSave();
        expect(onDelete).toBeCalled();
      });

      it('calls onClose', () => {
        const w = wrapper();

        w.setState({ value: 'note' });
        w.instance().handleSave();
        expect(onClose).toBeCalled();
      });
    });

    describe('type change', () => {
      it('calls onTypeChange',  () => {
        const w = wrapper();

        w.instance().handleTypeChange();
        expect(onTypeChange).toBeCalledWith('note');
      });

      it('calls onTypeChange with changed value',  () => {
        const w = wrapper();

        w.setState({ value: 'good' });
        w.instance().handleTypeChange();
        expect(onTypeChange).toBeCalledWith('good');
      });

      it('calls onDelete', () => {
        const w = wrapper();

        w.setState({ value: '' });
        w.instance().handleTypeChange();
        expect(onDelete).toBeCalled();
      });
    });
  });
});
