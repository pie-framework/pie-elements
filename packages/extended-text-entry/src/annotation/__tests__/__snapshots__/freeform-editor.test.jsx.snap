// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`freeform editor snapshots opened 1`] = `
<WithStyles(Popover)
  anchorEl={Object {}}
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  classes={
    Object {
      "paper": "positive",
    }
  }
  elevation={2}
  onClose={[Function]}
  open={true}
  style={
    Object {
      "marginTop": "10px",
      "transition": "margin-top 2s ease-out",
    }
  }
  transformOrigin={
    Object {
      "horizontal": "left",
      "vertical": "top",
    }
  }
>
  <div
    className="positive"
  >
    <TextField
      InputProps={
        Object {
          "disableUnderline": true,
        }
      }
      autoFocus={true}
      id="annotation-editor"
      multiline={true}
      onChange={[Function]}
      required={false}
      rows={1}
      rowsMax={4}
      select={false}
      style={
        Object {
          "padding": "2px 5px",
          "width": "95%",
        }
      }
      value="note"
      variant="standard"
    />
    <div>
      <div
        onClick={[MockFunction]}
      >
        Delete
      </div>
      <div
        className="positive"
        onClick={[Function]}
      >
        Pink
      </div>
      <div
        onClick={[Function]}
      >
        Save
      </div>
    </div>
  </div>
</WithStyles(Popover)>
`;

exports[`freeform editor snapshots opened, negative type 1`] = `
<WithStyles(Popover)
  anchorEl={Object {}}
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  classes={
    Object {
      "paper": "negative",
    }
  }
  elevation={2}
  onClose={[Function]}
  open={true}
  style={
    Object {
      "marginTop": "10px",
      "transition": "margin-top 2s ease-out",
    }
  }
  transformOrigin={
    Object {
      "horizontal": "left",
      "vertical": "top",
    }
  }
>
  <div
    className="negative"
  >
    <TextField
      InputProps={
        Object {
          "disableUnderline": true,
        }
      }
      autoFocus={true}
      id="annotation-editor"
      multiline={true}
      onChange={[Function]}
      required={false}
      rows={1}
      rowsMax={4}
      select={false}
      style={
        Object {
          "padding": "2px 5px",
          "width": "95%",
        }
      }
      value="note"
      variant="standard"
    />
    <div>
      <div
        onClick={[MockFunction]}
      >
        Delete
      </div>
      <div
        className="negative"
        onClick={[Function]}
      >
        Green
      </div>
      <div
        onClick={[Function]}
      >
        Save
      </div>
    </div>
  </div>
</WithStyles(Popover)>
`;

exports[`freeform editor snapshots renders 1`] = `
<WithStyles(Popover)
  anchorEl={Object {}}
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  classes={
    Object {
      "paper": "positive",
    }
  }
  elevation={2}
  onClose={[Function]}
  open={false}
  style={
    Object {
      "marginTop": "10px",
      "transition": "margin-top 2s ease-out",
    }
  }
  transformOrigin={
    Object {
      "horizontal": "left",
      "vertical": "top",
    }
  }
>
  <div
    className="positive"
  >
    <TextField
      InputProps={
        Object {
          "disableUnderline": true,
        }
      }
      autoFocus={true}
      id="annotation-editor"
      multiline={true}
      onChange={[Function]}
      required={false}
      rows={1}
      rowsMax={4}
      select={false}
      style={
        Object {
          "padding": "2px 5px",
          "width": "95%",
        }
      }
      value="note"
      variant="standard"
    />
    <div>
      <div
        onClick={[MockFunction]}
      >
        Delete
      </div>
      <div
        className="positive"
        onClick={[Function]}
      >
        Pink
      </div>
      <div
        onClick={[Function]}
      >
        Save
      </div>
    </div>
  </div>
</WithStyles(Popover)>
`;
